src/statemachine.rs

```diff
@@ -683,50 +724,95 @@ impl StateMachineCtap2 {
+        );
 
-                info!("Device {:?} continues with the reset process", dev.id());
-                debug!("------------------------------------------------------------------");
-                debug!("{:?}", reset);
-                debug!("------------------------------------------------------------------");
+        self.transaction = Some(try_or!(transaction, move |e| cbc.call(Err(e))));
+    }
 
-                let resp = dev.send_cbor(&reset);
-                if resp.is_ok() {
-                    send_status(
-                        &status,
-                        crate::StatusUpdate::Success {
-                            dev_info: dev.get_device_info(),
-                        },
-                    );
-                    // The DeviceSelector could already be dead, but it might also wait
-                    // for us to respond, in order to cancel all other tokens in case
-                    // we skipped the "blinking"-action and went straight for the actual
-                    // request.
-                    let _ = selector.send(DeviceSelectorEvent::SelectedToken(dev.id()));
-                }
+    pub fn set_or_change_pin_helper(
+        dev: &mut Device,
+        mut current_pin: Option<Pin>,
+        new_pin: Pin,
+        status: Sender<crate::StatusUpdate>,
+        callback: StateCallback<crate::Result<crate::ResetResult>>,
+    ) {
+        let (mut shared_secret, authinfo) = match dev.establish_shared_secret() {
+            Ok(s) => s,
+            Err(e) => {
+                callback.call(Err(AuthenticatorError::HIDError(e)));
+                return;
+            }
+        };
 
-                match resp {
-                    Ok(()) => callback.call(Ok(())),
-                    Err(HIDError::DeviceNotSupported) | Err(HIDError::UnsupportedCommand) => {}
-                    Err(HIDError::Command(CommandError::StatusCode(
-                        StatusCode::ChannelBusy,
-                        _,
-                    ))) => {}
-                    Err(e) => {
-                        warn!("error happened: {}", e);
-                        callback.call(Err(AuthenticatorError::HIDError(e)));
-                    }
+        // With CTAP2.1 we will have an adjustable required length for PINs
+        if new_pin.as_bytes().len() < 4 {
+            callback.call(Err(AuthenticatorError::PinError(PinError::PinIsTooShort)));
+            return;
+        }
+
+        if new_pin.as_bytes().len() > 64 {
+            callback.call(Err(AuthenticatorError::PinError(PinError::PinIsTooLong(
+                new_pin.as_bytes().len(),
+            ))));
+            return;
+        }
+
+        // Check if a client-pin is already set, or if a new one should be created
+        let res = if authinfo.options.client_pin.unwrap_or_default() {
+            let mut res;
+            let mut error = PinError::PinRequired;
+            loop {
+                let curr_pin = match current_pin.take() {
```

Comment:
This is so it works for both non-interactive and interactive cases.

In the non-interactive case, `current_pin` is always `None` and we will ask the user every time via StatusUpdates for a PIN.

For the interactive-case (`manage()`), I want to be able to hand in a PIN from the start, to avoid the Ping-Pong of getting the device. But in case this handed in PIN is wrong, I don't want to run into an endless-loop and lock the device (because it would retry that PIN over and over). 
So I `take()` the handed in PIN. If that is wrong, there will be a fallback to "asking the user" (which the about-page currently doesn't utilize, but simply lets the interaction fail, so the user can try again with the HTML-form).

I'll add an explanatory comment.

Reply:
I mean, yes we could probably move that out of the loop, but at the cost of code duplication and a generally longer function (or additional refactoring).

Can do so, if you wish.